forest/bonsai/simulate_gps_data.py

```diff
@@ -1117,31 +1122,300 @@ def process_switches(
-                s_lower = s+i*24*60*60*1000+j*60*60*1000
-                s_upper = s+i*24*60*60*1000+(j+1)*60*60*1000
-                temp = obs_pd[(obs_pd["timestamp"]>=s_lower)&(obs_pd["timestamp"]<s_upper)]
-                [y,m,d,h,mins,sec] = stamp2datetime(s_lower/1000,"UTC")
-                filename = f"{y}-{m:0>2}-{d:0>2}-{h:0>2}_00_00.csv"
-                temp.to_csv(data_folder+"/user_"+str(user+1)+"/gps/"+filename,index = False)
+def load_attributes(
+    attributes: Dict[str, Dict],
+) -> Tuple[Dict[int, Attributes], Dict[int, Dict[str, int]]]:
+    """Loads the attributes of each person.
+
+    Args:
+        attributes: Dictionary of attributes of each person,
+            loaded from json file.
+    Returns:
+        attributes: (dictionary) contains attributes of each person,
+            loaded from json file.
+        switches: (dictionary) contains possible changes of attributes
+            in between of simulation
+    Raises:
+        ValueError: if the format of the json file is not correct.
+    """
+
+    attributes_dictionary: Dict[int, Attributes] = {}
+    switches_dictionary: Dict[int, Dict[str, int]] = {}
+
+    for key in attributes.keys():
+        match = re.search(r"[0-9]*-?[0-9]+", key)
+        if match is None:
+            raise ValueError(f"Wrong format in attributes.json on {key}")
+        users = match.group(0).split("-")
+        if len(users) == 1:
+            up_to = int(users[0])
+        else:
+            up_to = int(users[1])
+        for user in range(int(users[0]), up_to + 1):
+            attrs = Attributes(**attributes[key])
+            switches = process_switches(attributes, key)
+            attributes_dictionary[user] = attrs
+            switches_dictionary[user] = switches
+
+    return attributes_dictionary, switches_dictionary
+
+
+def generate_addresses(country: str, city: str) -> np.ndarray:
+    """Generates multiple addresses.
+
+    Args:
+        country: (str) country of the persons
+        city: (str) city of the persons
+    Returns:
+        addresses: (np.ndarray) contains address coordinates of each person
+    Raises:
+        RuntimeError: if the api raises error for too many tries
+        ValueError: if the api returns no results
+    """
+
+    overpy_query = f"""
+    [out:json];
+    area["ISO3166-1"="{country}"][admin_level=2] -> .country;
+    area["name"="{city}"] -> .city;
+    node(area.country)(area.city)["addr:street"];
+    out center 150;
+    """
+
+    response = requests.get(
+            OVERPASS_URL,
+            params={"data": overpy_query}, timeout=5 * 60
+        )
+    if response.status_code != 200:
+        raise RuntimeError(
+            f"Overpass API returned error {response.status_code}"
+            )
+
+    res = response.json()
+    try:
+        index = np.random.choice(
+            range(len(res["elements"])), 100, replace=False
+        )
+    except ValueError:
+        raise ValueError(
+            """
```

Comment:
It's better to use implicit string concatenation here instead of multiline strings because they contain `\n` which affect output formatting.